app/models/pacbio/plate_creator.rb

```diff
@@ -0,0 +1,237 @@
+# frozen_string_literal: true
+
+module Pacbio
+  # PlateCreator
+  # This will build a plate with wells and samples
+  # we could use the respective models but that would cause more complexity
+  # It is done in a cascading manner
+  # Assumptions are:
+  #  *the plate will probably have an external barcode to map to the LIMS it has originated from
+  #  *if no barcode then a TRAC one is generated
+  #  *the plate will have one or more wells
+  #  *each well can have a sample but it is not required
+  class PlateCreator
+    include ActiveModel::Model
+
+    attr_reader :plate_wrappers
+
+    validate :check_plates
+
+    def initialize(attributes = {})
+      @plate_wrappers = attributes[:plates].try(:collect) { |plate| PlateWrapper.new(plate) }
+    end
+
+    # Previously checking validity and then saving but this causes problems because
+    # things are validated several times so it is inefficient
+    # So do save bang and if anything is not valid it will raise an error
+    # We can then collect those errors on rescue
+    # @return [Boolean] was everything saved correctly
+    def save!
+      ActiveRecord::Base.transaction do
+        raise ActiveRecord::RecordInvalid if plate_wrappers.nil? || plate_wrappers.try(:empty?)
+
+        plate_wrappers.collect(&:save!)
+        true
+      end
+    rescue ActiveRecord::RecordInvalid
+      # we need to cascade the errors up to the current error object
+      # otherwise it will look like there are no errors
+      check_plates
+      false
+    end
+
+    # useful for returning in an api response
+    def plates
+      @plates ||= plate_wrappers.collect(&:plate)
+    end
+
+    def check_plates
+      if plate_wrappers.empty?
+        errors.add(:plates, 'should be present')
+        return
+      end
+
+      plate_wrappers.each do |plate_wrapper|
+        next if plate_wrapper.valid?
+
+        plate_wrapper.errors.each do |k, v|
+          errors.add(k, v)
+        end
+      end
+    end
+
+    # PlateWrapper
+    # creates a standard Plate
+    # creates an array of wells which are instances of WellWrapper
+    # A plate will always be valid as a barcode can be generated
+    # There needs to be some wells
+    # If the wells are present they will need to be valid
+    class PlateWrapper
+      include ActiveModel::Model
+
+      attr_reader :well_wrappers, :plate
+
+      delegate :barcode, to: :plate
+
+      validate :check_wells
+
+      # @param attributes [Hash] hash of attributes
+      # this will contain:
+      #  * barcode
+      #  * array of wells
+      def initialize(attributes = {})
+        @plate = ::Plate.new(attributes.except(:wells))
+        @well_wrappers = attributes[:wells].try(:collect) do |well|
+          WellWrapper.new(well.merge(plate: plate))
+        end
+      end
+
+      def save!
+        plate.save!
+        raise ActiveRecord::RecordInvalid if well_wrappers.nil? || well_wrappers.try(:empty?)
```

Comment:
could this be changed to just 'raise ActiveRecord::RecordInvalid unless valid?' or if avoiding adding error messaging does well_wrappers.blank? accomplish the same thing